partition.go

```diff
@@ -0,0 +1,141 @@
+package schema
+
+import (
+	"encoding/binary"
+	"hash/fnv"
+	"sort"
+
+	"github.com/cespare/xxhash"
+	jump "github.com/dgryski/go-jump"
+)
+
+type PartitionByMethod uint8
+
+const (
+	PartitionByOrg PartitionByMethod = iota
+	PartitionBySeries
+	PartitionBySeriesWithTags
+	PartitionBySeriesWithTagsFnv
+)
+
+func (m *MetricData) PartitionID(method PartitionByMethod, partitions int32) (int32, error) {
+	var partition int32
+
+	switch method {
+	case PartitionByOrg:
+		h := fnv.New32a()
+		err := binary.Write(h, binary.LittleEndian, uint32(m.OrgId))
+		if err != nil {
+			return 0, err
+		}
+		partition = int32(h.Sum32()) % partitions
+		if partition < 0 {
+			partition = -partition
+		}
+	case PartitionBySeries:
+		h := fnv.New32a()
+		h.Write([]byte(m.Name))
+		partition = int32(h.Sum32()) % partitions
+		if partition < 0 {
+			partition = -partition
+		}
+	case PartitionBySeriesWithTags:
+		h := xxhash.New()
+		h.WriteString(m.Name)
+		sort.Strings(m.Tags)
+		for _, t := range m.Tags {
+			if len(t) >= 5 && t[:5] == "name=" {
+				continue
+			}
+
+			h.WriteString(";")
+			h.WriteString(t)
+		}
+		partition = jump.Hash(h.Sum64(), int(partitions))
+	case PartitionBySeriesWithTagsFnv:
+		h := fnv.New32a()
+		h.Write([]byte(m.Name))
+		sort.Strings(m.Tags)
+		for _, t := range m.Tags {
+			if len(t) >= 5 && t[:5] == "name=" {
+				continue
+			}
+
+			h.Write([]byte(";"))
+			h.Write([]byte(t))
+		}
+		partition = int32(h.Sum32()) % partitions
+		if partition < 0 {
+			partition = -partition
+		}
+	default:
+		return 0, ErrUnknownPartitionMethod
+	}
+
+	return partition, nil
+}
+
+func (m *MetricDefinition) PartitionID(method PartitionByMethod, partitions int32) (int32, error) {
+	var partition int32
+
+	switch method {
+	case PartitionByOrg:
+		h := fnv.New32a()
+		err := binary.Write(h, binary.LittleEndian, uint32(m.OrgId))
+		if err != nil {
+			return 0, err
+		}
+		partition = int32(h.Sum32()) % partitions
+		if partition < 0 {
+			partition = -partition
+		}
+	case PartitionBySeries:
+		h := fnv.New32a()
+		h.Write([]byte(m.Name))
+		partition = int32(h.Sum32()) % partitions
+		if partition < 0 {
+			partition = -partition
+		}
+	case PartitionBySeriesWithTags:
+		h := xxhash.New()
+		if len(m.nameWithTags) > 0 {
```

Comment:
Is it faster to check `len(m.nameWithTags) > 0` or `m.nameWithTags != ""`?

Reply:
instead of generating the string this could just us `m.NameWithTags()`

Reply:
+1 what replay said

Reply:
so i just made the change...
```
-               if len(m.nameWithTags) > 0 {
-                       h.WriteString(m.nameWithTags)
-               } else {
-                       if err := writeSortedTagString(h, m.Name, m.Tags); err != nil {
-                               return 0, err
-                       }
-               }
+               h.WriteString(m.NameWithTags())
```

note that if nameWithTags is not set, it will now create the buffer, convert to string, call WriteString which turns the string back into []byte using some unsafe and reflect.
arguably this is some overhead, but on the plus side, it has the side effect of setting the nameWithTags property. which is probably desirable.